source/FormationPattern.cpp

```diff
@@ -15,44 +15,128 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 #include "FormationPattern.h"
 
+#include "Angle.h"
 #include "DataNode.h"
 
 #include <cmath>
 
 using namespace std;
 
-namespace {
-	const Point defaultFormationPoint = Point();
-}
-
 
-
-FormationPattern::PositionIterator::PositionIterator(const FormationPattern &pattern)
-	: pattern(pattern)
+FormationPattern::PositionIterator::PositionIterator(const FormationPattern &pattern,
+		double diameterToPx, double widthToPx, double heightToPx, double centerBodyRadius,
+		unsigned int shipsToPlace)
+	: pattern(pattern), shipsToPlace(shipsToPlace), centerBodyRadius(centerBodyRadius),
+		diameterToPx(diameterToPx), widthToPx(widthToPx), heightToPx(heightToPx)
 {
-	positionIt = pattern.positions.begin();
+	MoveToValidPosition();
 }
 
 
 
 const Point &FormationPattern::PositionIterator::operator*()
 {
-	if(positionIt == pattern.positions.end())
-		return defaultFormationPoint;
-	return *positionIt;
+	return currentPoint;
 }
 
 
 
 FormationPattern::PositionIterator &FormationPattern::PositionIterator::operator++()
 {
-	if(positionIt != pattern.positions.end())
-		++positionIt;
+	if(!atEnd)
+	{
+		position++;
+		MoveToValidPosition();
+	}
+	// Skip positions too close to the center body
+	while(!atEnd && currentPoint.Length() <= centerBodyRadius)
+	{
+		position++;
+		MoveToValidPosition();
+	}
+
+	// Number of ships is used as number of remaining ships still to be placed.
+	if(shipsToPlace > 0)
+		--shipsToPlace;
 	return *this;
 }
 
 
 
+void FormationPattern::PositionIterator::MoveToValidPosition()
+{
+	// If we cannot calculate any new positions, then just return center point.
+	if(atEnd)
+	{
+		currentPoint = Point();
+		return;
+	}
+
+	// Check if there are any lines available.
+	unsigned int lines = pattern.Lines();
+	if(lines < 1)
```

Comment:
we can return at this point, you could even merge this with the previous if

Reply:
Good idea.